### ISSUE TITLE
WebGL errors in many webxr samples

### ISSUE THREAD

**Issue by Manishearth** [servo/servo#24373]
```
RUST: ERROR - canvas::webgl_thread - Last GL operation failed: BindTexture(3553, Some(WebGLTextureId(14)))
thread '<unnamed>' panicked at 'assertion failed: `(left == right)`
  left: `1286`,
 right: `0`: Unexpected WebGL error: 0x506 (1286)', components\canvas\webgl_thread.rs:1654:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

I keep getting this in many three.js examples (webvr_ballshooter.html, webvr_sculpt.html are some) on the hololens

I can also trigger it in https://manishearth.github.io/webgl-to-webxr/webxr-input.html if I trigger an exception near the `let input_pose`. If you change that line to throw it should happen.

cc @asajeffrey

**Comment by asajeffrey** [comment]
Lets see if this magically goes away once surfman is in place.

**Comment by asajeffrey** [comment]
Once servo+surfman is ready, WebXRGLLayer will have a completely different code path for creating the framebuffer, since it'll be backed by a swap chain rather than a GL framebuffer.

**Comment by jdm** [comment]
I understand. I think it's still valuable to address an issue that's preventing experimenting with existing XR content rather than waiting until surfman is ready, though.

**Comment by jdm** [comment]
This is fascinating - https://manishearth.github.io/webgl-to-webxr/webxr-broken.js triggers this panic, and https://manishearth.github.io/webgl-to-webxr/webxr-ar.html does not. The only difference is that webxr-broken.js [throws an exception](https://github.com/Manishearth/webgl-to-webxr/blob/d1c7f2e2d564a70c2025627a8f19932b0b00aa80/webxr-broken.js#L219), while the GL error occurs when running the WebGL commands stemming from [creating the XRWebGLLayer](https://github.com/Manishearth/webgl-to-webxr/blob/d1c7f2e2d564a70c2025627a8f19932b0b00aa80/webxr-broken.js#L112). This suggests that we're running a XR rAF callback before we've actually completely processed the operations that go into creating the XR webgl layer.

**Comment by jdm** [comment]
I suspect the difference in behaviour comes from not running `gl.finish()` in the VR callback when the exception occurs. I'm seeing if calling Finish before returning from the XRWebGLLayer constructor is enough to avoid this problem.

**Comment by asajeffrey** [comment]
One possible root cause for this is that we assert (in debug mode) that there's been no GL errors, but I suspect in some cases that relies on the WebGL code being correct. For example `glClear` can error in all sorts of ways, some of which depend on the WebGL code being executed. AFAICT if the WebGL code is incorrect, then rather than reporting a WebGL error, we're panicing at https://github.com/servo/servo/blob/3d529b1b8859cbb5f9f8337ec8343c3254fd4224/components/canvas/webgl_thread.rs#L1678-L1684

**Comment by jdm** [comment]
Sadly I can't reproduce this under the glwindow backend on mac.

**Comment by jdm** [comment]
Aha, I tracked this down by setting a breakpoint in ANGLE's Context::handleError. This showed me that the GL API that triggers the error is a call to ReadPixels that originates from the webxr openxr backend in render_animation_frame. At this point, the webgl thread has bound the XR opaque framebuffer's texture but hasn't executed the remaining commands that make up the XR webgl layer's initialization. ReadPixels fails because the framebuffer's texture isn't ready yet, and it leaves a behind GL error that is incorrectly attributed to the subsequent TexImage2D call on the WebGL thread.

So ultimately we are in the situation that I suspected - the XR animation frame callback is executing before the XR webgl layer is fully initialized, and due to the current readback situation this means we execute some GL in an unanticipated state. #24408 avoids this problem because it forces a synchronization point, so the code to kick off the XR raF doesn't start running until the XR webgl layer is fully initialized in the webgl thread.
